net/fetcher.go

```diff
@@ -22,6 +30,212 @@ type Fetcher interface {
 	FetchTipSets(ctx context.Context, tsKey types.TipSetKey, from peer.ID, done func(ts types.TipSet) (bool, error)) ([]types.TipSet, error)
 }
 
+// GraphSyncFetcher is used to fetch data over the network.  It is implemented with
+// a persistent bitswap session on a networked blockservice.
+type GraphSyncFetcher struct {
+	gs        *graphsync.GraphSync
+	validator consensus.BlockSyntaxValidator
+	bridge    ipldbridge.IPLDBridge
+	bs        bstore.Blockstore
+}
+
+// NewGraphSyncFetcher returns a BitswapFetcher wired up to the input BlockService and a newly
+// initialized persistent session of the block service.
+func NewGraphSyncFetcher(ctx context.Context, network gsnet.GraphSyncNetwork, bridge ipldbridge.IPLDBridge, blockstore bstore.Blockstore,
+	bv consensus.BlockSyntaxValidator) *GraphSyncFetcher {
+	gsf := &GraphSyncFetcher{
+		bs:        blockstore,
+		validator: bv,
+		bridge:    bridge,
+	}
+	gsf.gs = graphsync.New(ctx, network, bridge, gsf.loader, gsf.storer)
```

Comment:
Blocking: this method of construction makes it very difficult to unit test the fetcher, since a fake graphsync cannot be injected. Please re-arrange construction so that we can exhaustively test the fetcher's behaviour and response to various things graphsync could return, like failures and partial responses, via a fake implementation of a consumer-defined interface for the fetcher's needs from graphsync.

I can see that the store interface methods are going to make that difficult, and you'll need to couple the store to the graphsync before wrapping it in a fetcher.

Reply:
The good news is the next step -- I don't think the loader/storer should be in fetcher nor should the bridge or the network. The method on the bridge is just a convenience method and probably not needed. I posted the PR mostly cause this is the first time I got a test to pass.